File: domain/src/main/java/com/beat/domain/promotion/dao/PromotionRepository.java
package com.beat.domain.promotion.dao;

import com.beat.domain.promotion.domain.CarouselNumber;
import com.beat.domain.promotion.domain.Promotion;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

public interface PromotionRepository extends JpaRepository<Promotion, Long> {
	List<Promotion> findAll();

	@Modifying(clearAutomatically = true)
	@Transactional
	@Query("DELETE FROM Promotion p WHERE p.id IN :promotionIds")
	void deleteByPromotionIds(@Param("promotionIds") List<Long> promotionIds);

	@Query("SELECT p FROM Promotion p WHERE p.carouselNumber = :carouselNumber")
	Optional<Promotion> findByCarouselNumber(@Param("carouselNumber") CarouselNumber carouselNumber);
}

File: core/domain/src/main/kotlin/com/beat/domain/promotion/repository/PromotionRepository.kt
package com.beat.domain.promotion.repository

import com.beat.domain.promotion.model.CarouselNumber
import com.beat.domain.promotion.model.Promotion
import java.util.*

@JvmSuppressWildcards
interface PromotionRepository {
    fun findAll(): List<Promotion>

    fun findById(promotionId: Long?): Optional<Promotion>

    fun save(promotion: Promotion): Promotion

    fun saveAll(promotions: List<Promotion>): List<Promotion>

    fun deleteByPromotionIds(promotionIds: List<Long>)

    fun deleteByPerformanceId(performanceId: Long?)

    fun findByCarouselNumber(carouselNumber: CarouselNumber): Optional<Promotion>
}
